main/tests/server/src/com/google/refine/commands/row/AddRowsCommandTests.java

```diff
@@ -0,0 +1,210 @@
+
+import java.io.IOException;
+import java.io.Serializable;
+
+import javax.servlet.ServletException;
+
+import com.fasterxml.jackson.core.JsonParseException;
+import com.fasterxml.jackson.databind.JsonNode;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import com.google.refine.commands.Command;
+import com.google.refine.commands.CommandTestBase;
+import com.google.refine.model.Project;
+import com.google.refine.util.ParsingUtilities;
+
+public class AddRowsCommandTests extends CommandTestBase {
+
+    private Project project = null;
+    private String[] additionalRows = {
+            "{ starred: false, flagged: false, cells: []}",
+            "{ starred: false, flagged: false, cells: []}",
+    };
+
+    protected AddRowsCommand command;
+
+    @BeforeMethod
+    public void setUpCommand() {
+        command = new AddRowsCommand();
+        project = createProject(
+                new String[] { "Column 1", "Column 2" },
+                new Serializable[][] {
+                        { "c", "3" },
+                        { "d", "4" },
+                });
+        when(request.getParameter("project")).thenReturn(Long.toString(project.id));
+    }
+
+    @Test
+    // If CSRF token is missing, respond with CSRF error
+    public void testCSRFProtection() throws ServletException, IOException {
+        command.doPost(request, response);
+        assertCSRFCheckFailed();
+    }
+
+    @Test
+    // If successful request, respond with "Code" equals "ok"
+    public void testSuccessResponseCode() throws ServletException, IOException {
+        when(request.getParameter("csrf_token")).thenReturn(Command.csrfFactory.getFreshToken());
+        when(request.getParameter(AddRowsCommand.INDEX_PARAMETER)).thenReturn("0");
+        when(request.getParameterValues(AddRowsCommand.ROWS_PARAMETER)).thenReturn(additionalRows);
+
+        command.doPost(request, response);
+
+        JsonNode node = ParsingUtilities.mapper.readValue(writer.toString(), JsonNode.class);
+        assertEquals(node.get("code").toString(), "\"ok\"");
+    }
+
+    @Test
+    // If successful request, responses contains complete HistoryEntry object
+    public void testSuccessResponseHistoryEntrySchema() throws ServletException, IOException {
+        when(request.getParameter("csrf_token")).thenReturn(Command.csrfFactory.getFreshToken());
+        when(request.getParameter(AddRowsCommand.INDEX_PARAMETER)).thenReturn("0");
+        when(request.getParameterValues(AddRowsCommand.ROWS_PARAMETER)).thenReturn(additionalRows);
+
+        command.doPost(request, response);
+
+        JsonNode node = ParsingUtilities.mapper.readValue(writer.toString(), JsonNode.class);
+        assertNotNull(node.get("historyEntry"));
+        assertNotNull(node.get("historyEntry").get("id"));
+        assertNotNull(node.get("historyEntry").get("description"));
+        assertNotNull(node.get("historyEntry").get("time"));
```

Comment:
Would it make sense to merge the two tests `testSuccessResponseCode` and `testSuccessResponseHistoryEntrySchema` since they seem to be making different assertions in the same testing scenario? You could move your assertions about `historyEntry` to the first test above. Or is it intentional that they are separate?

Reply:
I see what you mean. Ok I'll merge those to one test to cover the response body of a successful response.

I initially separated them because I wanted `testSuccessResponseCode` to test the HTTP response status code, after past confusion about status codes (#5772). But, the mock for HttpServletResponse doesn't provide a real status code.